pkg/xds/server/callbacks/dataplane_callbacks.go

```diff
@@ -0,0 +1,154 @@
+	// OnFirstStreamConnected is similar to OnStreamConnected but if there are many xDS streams, it is only executed for the first stream.
+	OnFirstStreamConnected(streamID core_xds.StreamID, dpKey core_model.ResourceKey, ctx context.Context, metadata core_xds.DataplaneMetadata) error
+
+	// OnStreamDisconnected is executed when Dataplane stream disconnects.
+	OnStreamDisconnected(streamID core_xds.StreamID, dpKey core_model.ResourceKey)
+	// OnLastStreamDisconnected is executed only when the last stream of Dataplane disconnects.
+	OnLastStreamDisconnected(streamID core_xds.StreamID, dpKey core_model.ResourceKey)
+}
+
+type dpCallbacks struct {
+	callbacks DataplaneCallbacks
+	util_xds.NoopCallbacks
+
+	sync.RWMutex
+	dpStreams     map[core_xds.StreamID]dpStream
+	activeStreams map[core_model.ResourceKey]int
+}
+
+func DataplaneCallbacksToXdsCallbacks(callbacks DataplaneCallbacks) util_xds.Callbacks {
+	return &dpCallbacks{
+		callbacks:     callbacks,
+		dpStreams:     map[core_xds.StreamID]dpStream{},
+		activeStreams: map[core_model.ResourceKey]int{},
+	}
+}
+
+type dpStream struct {
+	dp  *core_model.ResourceKey
+	ctx context.Context
+}
+
+var _ util_xds.Callbacks = &dpCallbacks{}
+
+func (d *dpCallbacks) OnStreamClosed(streamID core_xds.StreamID) {
+	d.Lock()
+	defer d.Unlock()
+	dpStream := d.dpStreams[streamID]
+	if dpKey := dpStream.dp; dpKey != nil {
+		d.activeStreams[*dpKey]--
+		d.callbacks.OnStreamDisconnected(streamID, *dpKey)
+		if d.activeStreams[*dpKey] == 0 {
+			d.callbacks.OnLastStreamDisconnected(streamID, *dpKey)
+			delete(d.activeStreams, *dpKey)
+		}
+	}
+	delete(d.dpStreams, streamID)
+}
+
+func (d *dpCallbacks) OnStreamRequest(streamID core_xds.StreamID, request util_xds.DiscoveryRequest) error {
+	if request.NodeId() == "" {
+		// from https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack-nack-and-versioning:
+		// Only the first request on a stream is guaranteed to carry the node identifier.
+		// The subsequent discovery requests on the same stream may carry an empty node identifier.
+		// This holds true regardless of the acceptance of the discovery responses on the same stream.
+		// The node identifier should always be identical if present more than once on the stream.
+		// It is sufficient to only check the first message for the node identifier as a result.
+		return nil
+	}
+
+	d.RLock()
+	alreadyProcessed := d.dpStreams[streamID].dp != nil
+	d.RUnlock()
+	if alreadyProcessed {
+		return nil
+	}
+
+	proxyId, err := core_xds.ParseProxyIdFromString(request.NodeId())
+	if err != nil {
+		return errors.Wrap(err, "invalid node ID")
+	}
+	dpKey := proxyId.ToResourceKey()
+	metadata := core_xds.DataplaneMetadataFromXdsMetadata(request.Metadata())
+	if metadata == nil {
+		return errors.New("metadata in xDS Node cannot be nil")
+	}
+
+	d.Lock()
+	dpStream := d.dpStreams[streamID]
+	dpStream.dp = &dpKey
+	d.dpStreams[streamID] = dpStream
```

Comment:
Interesting, I didn't know you can't do `d.dpStreams[streamID].dp = &dpKey` in GoLang, but that's of course makes sense :)